modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java

```diff
@@ -0,0 +1,299 @@
+        this.clusterService = clusterService;
+        this.placementDriver = placementDriver;
+        this.clock = clock;
+
+        addListeners();
+    }
+
+    @Override
+    public void close() {
+        if (!closeGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+    }
+
+    private void addListeners() {
+        catalogService.listen(CatalogEvent.INDEX_CREATE, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onIndexCreate(((CreateIndexEventParameters) parameters)).thenApply(unused -> false);
+        });
+
+        catalogService.listen(CatalogEvent.INDEX_DROP, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onIndexDrop(((DropIndexEventParameters) parameters)).thenApply(unused -> false);
+        });
+
+        placementDriver.listen(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onPrimaryReplicaElected(parameters).thenApply(unused -> false);
+        });
+    }
+
+    private CompletableFuture<?> onIndexCreate(CreateIndexEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            var startBuildIndexFutures = new ArrayList<CompletableFuture<?>>();
+
+            for (TablePartitionId primaryReplicaId : primaryReplicaIds) {
+                if (primaryReplicaId.tableId() == parameters.indexDescriptor().tableId()) {
+                    CompletableFuture<?> startBuildIndexFuture = getMvTableStorageFuture(parameters.causalityToken(), primaryReplicaId)
+                            .thenCompose(mvTableStorage -> awaitPrimaryReplicaFowNow(primaryReplicaId)
+                                    .thenAccept(replicaMeta -> tryStartBuildIndex(
+                                            primaryReplicaId,
+                                            parameters.indexDescriptor(),
+                                            mvTableStorage,
+                                            replicaMeta
+                                    ))
+                            );
+
+                    startBuildIndexFutures.add(startBuildIndexFuture);
+                }
+            }
+
+            return allOf(startBuildIndexFutures.toArray(CompletableFuture[]::new));
+        });
+    }
+
+    private CompletableFuture<?> onIndexDrop(DropIndexEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            indexBuilder.stopBuildingIndexes(parameters.indexId());
+
+            return completedFuture(null);
+        });
+    }
+
+    private CompletableFuture<?> onPrimaryReplicaElected(PrimaryReplicaEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
```

Comment:
If you are talking about `inBusyLockAsync`, it is called that because it returns the future, this is described in its documentation.

If you think it is worth improving the name or description of method `inBusyLockAsync`, then please create a ticket for this.

> Why do we even have a specific version for completable future

Because there are `inBusyLock` methods with the same semantics.

Reply:
I will explain my issue with the name. While reading the code, I assumed that passed closure will be executed in different thread. This changes the perception of the code completely. You start thinking that the operation is async, which implies different properties that you check. Why did you called a synchronous operation "Async", that's beyond me

Reply:
I can't find a convention that states that the code that returns the future must be executed in another thread.
Could you provide information where this is stated?

Reply:
> I can't find a convention that states that the code that returns the future must be executed in another thread.

There's no such convention. Never was. You are misinterpreting what I said.
I say that method, that has `Async` suffix, does execute the operation asynchronously. For example, `java.util.concurrent.CompletableFuture#runAsync(java.lang.Runnable)` runs a closure in a separate thread.
A better example is a pair `thenCompose` and `thenComposeAsync`. Both  these methods acept closure that returns the future, just like in your case. `thenCompose` executes the closure in the same thread, and `thenComposeAsync` executes closure in a threadpool.
In your case, you execute the closure in the same thread, but call the method `Async`. This is counter-intuitive.

> Could you provide information where this is stated?

I don't get it, what do I need to provide. This is what all people do without a formal written agreement, as far as I'm concerned.

Reply:
We discussed it personally until we found a better name for `inBusyLockAsync` so we’ll leave it as is.